game-core/src/main/java/games/strategy/engine/framework/startup/launcher/AbstractLauncher.java

```diff
@@ -2,52 +2,20 @@
 
 import java.awt.Component;
 
-import javax.swing.JOptionPane;
-import javax.swing.SwingUtilities;
-
-import games.strategy.engine.data.GameData;
-import games.strategy.engine.framework.startup.mc.GameSelectorModel;
-import games.strategy.engine.framework.ui.background.WaitWindow;
-
 /**
  * Abstract class for launching a game.
+ *
+ * @param <T> The type of object that gets returned by {@link #loadGame(Component)}
+ *        and is required by {@link #launchInternal(Component, Object)}.
  */
-public abstract class AbstractLauncher implements ILauncher {
-  protected final GameData gameData;
-  protected final GameSelectorModel gameSelectorModel;
-  protected final WaitWindow gameLoadingWindow;
-  protected final boolean headless;
-
-  protected AbstractLauncher(final GameSelectorModel gameSelectorModel) {
-    this(gameSelectorModel, false);
-  }
-
-  protected AbstractLauncher(final GameSelectorModel gameSelectorModel, final boolean headless) {
-    this.headless = headless;
-    if (this.headless) {
-      gameLoadingWindow = null;
-    } else {
-      gameLoadingWindow = new WaitWindow();
-    }
-    this.gameSelectorModel = gameSelectorModel;
-    gameData = gameSelectorModel.getGameData();
-  }
-
+public abstract class AbstractLauncher<T> implements ILauncher {
   @Override
   public void launch(final Component parent) {
-    if (!headless && !SwingUtilities.isEventDispatchThread()) {
-      throw new IllegalStateException("Wrong thread");
-    }
-    if (!headless && gameLoadingWindow != null) {
-      gameLoadingWindow.setLocationRelativeTo(JOptionPane.getFrameForComponent(parent));
-      gameLoadingWindow.setVisible(true);
-      gameLoadingWindow.showWait();
-    }
-    if (parent != null) {
-      JOptionPane.getFrameForComponent(parent).setVisible(false);
-    }
-    new Thread(() -> launchInNewThread(parent), "Triplea start thread").start();
+    final T result = loadGame(parent);
+    new Thread(() -> launchInternal(parent, result)).start();
```

Comment:
I really like seeing the method in this form. :+1: 

For a next PR: we can next inject the two abstract methods to avoid having an inheritance tree. Eg:
```
@Builder
@AllArgsConstructor
public abstract class Launcher<Optional<T>> implements ILauncher {
  
  private final Supplier<Optional<T>> gameLoader;
  private final BiConsumer<Component, T> gameLauncher;

 @Override
  public void launch(final Component parent) {
     gameLoader.get().ifPresent(game ->
          new Thread(() -> gameLoader.accept(parent, game)).start());
  }
}
``` 
The thing I like about the above is the loading and launching concerns become decoupled.

Reply:
Could you elaborate on that one?
I don't quite see the big difference on your approach vs the "current one" proposed in this PR.
It shouldn't make a difference declaring a method abstract vs passing it as a lambda?

Reply:
Sure, think strategy pattern. It's not necessarily important we inject lambdas, we are injecting interfaces and it is the interfaces that do the work, not the class implementation.

For example, consider the construction:
```
Launcher.builder()
    .gameLoader(someGameLoaderObject)
    .gameLauncher(someGameLauncherObject)
    .build();
```

From a testing perspective:
- To test the `launch` method of `Launcher`, we can inject two mock objects and verify things work accordingly. In turn we can test the launcher and loader objects independently. On the other hand with an inherited object, there are more dependencies when creating the object under test, it can be harder to test the loader or launcher on their own.


From a composition perspective:
- No inheritance hierarchy, can swap in and out just a loader or launcher implementation; inheritance makes this more difficult. We also avoid n-tier inheritance, currently it looks like `AbstractLauncher` has 'grand-children', changes to any class in the tree can impact others, it just overall reduces isolation and makes it harder to reason about the code.